File: MinimumInRotatedSortedArray.cpp
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes , took lots of time

// Your code here along with comments explaining your approach
//In a rotated sorted array, one side is always sorted at any given point. So the idea is to go to the unsorted side to find the minimum 
//We keep doing this, until our low pointer reaches the min.

class Solution {
public:
    int findMin(vector<int>& nums) 
    {
        int low =0;
        int high = nums.size()-1;
        while(low<high)
        {
            int mid = low + (high - low)/2;
            if(nums[mid]>nums[high])
                low = mid+1;
            else 
                high = mid;
        }
        return nums[low];
    }
};

File: Minimum_in_rotated_sorted_array.java
// Time Complexity :O(log n)
// Space Complexity : O(1) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : N/A


class Solution {
    public int findMin(int[] nums) {

     int l = 0;
     int r = nums.length -1;
     if (nums == null || nums.length == 0)
     return -1;
     
     while(l<=r){
        int mid = l + (r-l)/2;
        if(nums[l] <= nums[r]){return nums[l]; } 
        
        if( (mid == 0 || nums[mid] < nums[mid-1]) && 
        (mid == nums.length-1 || nums[mid] < nums[mid+1])){
        return nums[mid];
        } else if (nums[l] <= nums[mid]){
            l = mid + 1;
        }
        else{
            r = mid - 1;
        }
     }
    
    return 2829829;
    }
    
}
